### PR TITLE
Support for period (`.`) character in `Target.productName`

### PR DESCRIPTION
**Pull request by Lilfaen**
### Short description 📝

This adds period characters to the list of allowed characters for `productName`.

### Contributor checklist ✅

- [x] The code has been linted using run `./fourier lint tuist --fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Contributors have checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Review by danyf90 — APPROVED** [review]

**Review by kwridan — COMMENTED** [review]
Thanks for submitting this @Lilfaen 

Xcode does indeed permit `.` periods in the product and target names, however I believe this was omitted from Tuist as it can lead to some odd behaviours.

For example, if a target were to have the product name `app.core`, importing it from another module requires using the resolved module name `$(PRODUCT_NAME:c99extidentifier)` which does not permit `.` periods - in this example it would be `app_core`.

From my quick test in one of the fixtures, even attempting to use the resolved module name `app_core` yields in build failures.

Screenshot 2023-01-19 at 07 35 44

Mind sharing some more context on how you make use of product names with periods and how the import process works for those?

**Comment by Lilfaen** [comment]
Our project is (not yet) modular enough and the target whose product name contains the period is an app target, so we don't actually import in anywhere. The product name is shown inside the app, that's why we needed to include the period.

**Comment by kwridan** [comment]
Thanks for clarifying 👍 

> Our project is (not yet) modular enough and the target whose product name contains the period is an app target, so we don't actually import in anywhere. The product name is shown inside the app, that's why we needed to include the period.

In this case perhaps we could have one rule for application targets which does permits the `.` and the existing rule for everything else?

File: Tests/TuistGeneratorTests/Linter/TargetLinterTests.swift [hunk 1]
```diff
@@ -27,7 +27,7 @@ final class TargetLinterTests: TuistUnitTestCase {
             let target = Target.test(productName: productName)
             let got = self.subject.lint(target: target)
             let reason =
-                "Invalid product name '\(productName)'. This string must contain only alphanumeric (A-Z,a-z,0-9) and underscore (_) characters."
+                "Invalid product name '\(productName)'. This string must contain only alphanumeric (A-Z,a-z,0-9), period (.), and underscore (_) characters."
```
Comment (danyf90) [inline Tests/TuistGeneratorTests/Linter/TargetLinterTests.swift:30]: I guess you should remove the following line below:
```
        XCTAssertInvalidProductName("My.Framework")
```

**Comment by Lilfaen** [comment]
> In this case perhaps we could have one rule for application targets which does permits the `.` and the existing rule for everything else?

[Here's my idea how to achieve this. ](https://github.com/tuist/tuist/pull/4985/commits/9b48a67e6e1ce66edc7df2d9f8592c869cd9cb0a)

**Comment by kwridan** [comment]
> [Here's my idea how to achieve this.](https://github.com/tuist/tuist/pull/4985/commits/9b48a67e6e1ce66edc7df2d9f8592c869cd9cb0a)

we'll need to also condition the lint message 🤐 to avoid confusion

**Comment by Lilfaen** [comment]
I added a condition to the message and also modified the tests (thanks @danyf90)

File: Tests/TuistGeneratorTests/Linter/TargetLinterTests.swift [hunk 2]
```diff
@@ -23,26 +23,29 @@ final class TargetLinterTests: TuistUnitTestCase {
...
+        XCTAssertValidProductNameApp(Target.test(product: .app, productName: "MyApp.iOS"))
+        XCTAssertValidProductNameApp(Target.test(productName: "MyFramework_iOS"))
+        XCTAssertValidProductNameApp(Target.test(productName: "MyFramework"))
+        
+        XCTAssertInvalidProductNameApp(Target.test(product: .framework, productName: "MyFramework.iOS"))
+        XCTAssertInvalidProductNameApp(Target.test(productName: "MyFramework-iOS"))
+        XCTAssertInvalidProductNameApp(Target.test(productName: "ⅫFramework"))
+        XCTAssertInvalidProductNameApp(Target.test(productName: "ؼFramework"))
+        
+
```
Comment (danyf90) [inline Tests/TuistGeneratorTests/Linter/TargetLinterTests.swift:48]:
```suggestion
```

**Review by danyf90 — APPROVED** [review]

**Pull request merged**